cdn-video-server/src/toxiproxy/utils.ts

```diff
@@ -0,0 +1,155 @@
+import cp from "child_process";
+import path from "path";
+import {
+  IApiToxic,
+  INormalizedToxic,
+  IProxy,
+  ToxicType,
+} from "./toxiproxyTypes";
+
+// TODO Might be better written somewhere else
+const TOXIPROXY_CLI_EXEC = path.join(__dirname, "..", "..", "toxiproxy-cli");
+const CDN_PORT = 5000;
+const TOXIPROXY_PORT = 5001;
+
+function dictToString(dict: any): string {
+  return Object.entries(dict).reduce((
+    prev,
+    [key, value],
+  ) => `${prev}${key}=${value} `, "");
+}
+
+class Toxiproxy implements IProxy {
+  name: string = "CDN-ABR-PROXY";
+
+  toxics: Array<INormalizedToxic> = [];
+
+  listen: string;
+
+  upstream: string;
+
+  enabled: boolean = true;
+
+  constructor(
+    listen: string = `localhost:${TOXIPROXY_PORT}`,
+    upstream: string = `localhost:${CDN_PORT}`,
+  ) {
+    this.listen = listen;
+    this.upstream = upstream;
+  }
+
+  public async initialize(): Promise<void> {
+    return execProm(
+      `${TOXIPROXY_CLI_EXEC} create -l ${this.listen} -u ${this.upstream} ${this.name}`,
+    ).then(() => {});
+  }
+
+  public async updateToxics(toxics: IApiToxic[]): Promise<void> {
+    this.removeCurrentToxics();
+    const proms : Array<Promise<string>> = [];
+    for (let i = 0; i < toxics.length; i += 1) {
+      const toxic = toxics[i];
+      const normalizedToxic = normalizeToxic(toxic);
+      const attributeList = dictToString(normalizedToxic.attributes);
+      const streamFlag = normalizedToxic.stream === "downstream" ? "-d" : "-u";
+      proms.push(execProm(
+        `${TOXIPROXY_CLI_EXEC} toxic add -t ${normalizedToxic.type} -a ${attributeList} --tox ${normalizedToxic.toxicity} ${streamFlag} -n ${normalizedToxic.name} ${this.name}`,
+      ));
+      this.toxics.push(normalizedToxic);
+    }
+    return Promise.all(proms).then(() => {});
```

Comment:
We could simplify this by doing : 
```typescript
toxics.map(normalizeToxic).forEach(toxic => ...) 
```

Reply:
It's in no way objective, I just like the map syntax better than the loop syntax ^^